pkg/commands/artifact/sbom.go

```diff
@@ -0,0 +1,80 @@
+package artifact
+
+import (
+	"os"
+	"strings"
+
+	"github.com/aquasecurity/fanal/analyzer"
+	"github.com/aquasecurity/trivy/pkg/types"
+	"github.com/urfave/cli/v2"
+	"golang.org/x/xerrors"
+)
+
+type ArtifactType string
+
+const (
+	DockerImageArtifact   ArtifactType = "docker"
+	PackageArtifact       ArtifactType = "dir"
+	DockerArchiveArtifact ArtifactType = "archive"
+)
+
+var ArtifactTypes = []ArtifactType{DockerImageArtifact, PackageArtifact, DockerArchiveArtifact}
+
+var artifactTypeToScanner = map[ArtifactType]InitializeScanner{
+	DockerImageArtifact:   dockerScanner,
+	PackageArtifact:       filesystemScanner,
+	DockerArchiveArtifact: archiveScanner,
+}
+
+var artifactTypeToDisabledAnalyzers = map[ArtifactType][]analyzer.Type{
+	DockerImageArtifact:   analyzer.TypeLockfiles,
+	PackageArtifact:       analyzer.TypeIndividualPkgs,
+	DockerArchiveArtifact: analyzer.TypeLockfiles,
+}
+
+func analyzeArtifactPath(artifactPath string) (ArtifactType, string) {
+	// The user can specify the input artifact type by using the "<artifact-type>:" prefix before the artifact path
+	// e.g docker:ubuntu, package:/path/to/express
+	for _, artifactType := range ArtifactTypes {
+		prefix := string(artifactType) + ":"
+
+		if strings.HasPrefix(artifactPath, prefix) {
+			return artifactType, strings.Split(artifactPath, prefix)[1]
+		}
+	}
+
+	return detectArtifactType(artifactPath), artifactPath
```

Comment:
I'd like to return an error in this case. Like "invalid artifact type, "docker:", "dir:" or "artifact:" are allowed" or something like that.

Reply:
I don't want the artifact type prefix to be mandatory.
One of the goals of `trivy sbom` as a dedicated subcommand is to be able to determine the artifact type by itself.
I added the artifact type prefix for edge cases (for example if you want sbom for ubuntu image, but also have a local ubuntu directory, so you can specify exactly what you want trivy to generate sbom for)

Reply:
I think it is confusing and makes debugging harder for users. Filesystem scanning also supports a single file, so we don't know if it is an image archive or a dependency file. And if your local filesystem accidentally has the same file name as an image name you want to scan, the scanning will fail. I'm sure users will get confused.

Users must know what kind of artifact they want to scan. Adding a prefix is not a big effort. I feel that the disadvantages of the confusion without prefixes would be greater.

Reply:
I understand that it could be a little bit confusing, but it is a very uncommon edge case.
Most of the time this would not be a problem.

I think that the main goal of the dedicated sbom command is to let the user to just give the artifact path, and not specify it directly.

We can add logs that say:
"Searching for artifact <ARTIFACT_PATH>"
"Found 1 image "ARTIFACT" or "found 1 directory "ARTIFCT" or "found 1 archive "ARTIFACT"

Those logs can help the user understand which type of artifact was detected, and not spend time on debugging, but we also keep the user experience clean and nice by not requiring the user specify the artifact type.

I think we should cosider keeping the artifact type detection by trivy

Reply:
What is the benefit of removing prefixes for users? Do you think adding prefixes is a big effort? I think it is no big deal, so it is better to reduce confusion.

Reply:
I don't think that we need to remove the prefixes, just to make it optional 😁

I think the not requiring the prefixes simplifies the command for the users. In my opinion, this is a very important advantage, and the main goal of creating a dedicated subcommand for SBOM generation.

Another benefit of the "auto-detection", is that it makes the command generic, the user doesn't need to use different commands for different types, he has one unified command.
Then, if you use it in a CI pipeline for example, you can make it generic and you don't have to customize the command for each pipeline.

The use of a flag in a cli is to let the user customize the configuration of the command execution, like the SBOM format or quite mode. The artifact type is not up to the user to decide, it is a given thing. The user can't say that a directory is an image (for example).
The only reason I added the prefix option is to handle the spoken edge case - the user can use the prefixes to overcome it, but we want that by default he won't need it.

We want to let the user to generate SBOM in minimum effort, and I think the prefixes add unnecessary effort.

Reply:
>I don't think that we need to remove the prefixes, just to make it optional 😁

No, I mean users can remove prefixes (optional) and how important it is for users.

>In my opinion, this is a very important advantage

I'd like to know why it is very important.

>the main goal of creating a dedicated subcommand for SBOM generation.

As far as I understand, the main goal is generating multiple SBOM formats for multiple formats. The "auto-detection" sounds like another thing.

>Another benefit of the "auto-detection", is that it makes the command generic, the user doesn't need to use different commands for different types, he has one unified command.

I don't say we should not have `sbom` subcommand. We have one unified command regardless of prefixes.

>We want to let the user to generate SBOM in minimum effort, and I think the prefixes add unnecessary effort.

I'm asking how big the effort is. If users don't know what they want to scan, I agree we should detect it automatically. But they must know it. How hard do they specify it? I feel like the optional prefixes reduce very small effort and bring confusion. If you can tell me how hard it is, I'll be convinced🤔

Another idea is that the artifact type will be a container image by default and they can overwrite it with `fs` or others with the prefix.

Reply:
I don't think that container image as default is a good idea, it is more confusing.

I removed the auto-detection and updated the documentation - now it returns error when the artifact type prefix is unknown (or empty)

Reply:
I'm not saying the optional prefix is a bad idea, but we should sort out the pros and cons. Actually, I know it is optional in some tools, while required in other tools. We should discuss the trade-off. I'm curious why you think it is very important and we can make it optional if it makes sense.